### PR TITLE
refactor and move composer to status-im2

### PR DESCRIPTION
**Pull request by flexsurfer**
Composer has been fully reworked and simplified, also moved to status-im2

QA, please give it a good squeeze, especially on different Android devices

### CODE REVIEW

File: src/status_im2/contexts/chat/messages/composer/view.cljs [hunk 1]
```diff
@@ -0,0 +1,251 @@
+(ns status-im2.contexts.chat.messages.composer.view
```
Comment (flexsurfer) [inline src/status_im2/contexts/chat/messages/composer/view.cljs:1]: MOST IMPORTANT CHANGES ARE HERE

File: src/quo2/components/avatars/user_avatar.cljs [hunk 2]
```diff
@@ -134,7 +134,7 @@
       :style               {:width         outer-dimensions
                             :height        outer-dimensions
                             :border-radius outer-dimensions}}
-     (when (and ring? identicon?)
+     (when (and false (and ring? identicon?)) ;;TODO not implemented yet
```
Comment (ibrkhalil) [inline src/quo2/components/avatars/user_avatar.cljs:137]: This will always fail

Comment (J-Son89) [inline src/quo2/components/avatars/user_avatar.cljs:137]: can we add a corresponding issue for this?

File: src/status_im2/contexts/chat/messages/composer/view.cljs [hunk 3]
```diff
@@ -0,0 +1,250 @@
...
+        (set-bg-opacity 0)))))
+
+(defn get-input-content-change
+  [{:keys [chat-id] :as params}]
+  (fn [evt]
+    (let [new-height (oget evt "nativeEvent.contentSize.height")]
+      (swap! input/input-text-content-heights assoc chat-id new-height)
+      (if @initial-content-height
+        (do
+          ;;on Android when last symbol removed, height value is wrong, like 300px
```
Comment (J-Son89) [inline src/status_im2/contexts/chat/messages/composer/view.cljs:95]: what do you mean "like 300px"?? 🤔

File: src/quo2/components/avatars/user_avatar.cljs (hunk 2 above)
Comment (flexsurfer) [reply src/quo2/components/avatars/user_avatar.cljs:137]: i believe we have it

File: src/status_im2/contexts/chat/messages/composer/view.cljs (hunk 3 above)
Comment (flexsurfer) [reply src/status_im2/contexts/chat/messages/composer/view.cljs:95]: just big random number

File: src/quo2/components/avatars/user_avatar.cljs (hunk 2 above)
Comment (J-Son89) [reply src/quo2/components/avatars/user_avatar.cljs:137]: let's include the link here so :)

**Review by J-Son89 — APPROVED** [review]
I'm not so familiar with this code but everything else looks good to me 👍

File: src/status_im2/contexts/chat/messages/composer/controls/view.cljs [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+
+(defn view
+  [send-ref params insets chat-id images on-send]
+  [rn/view {:style (style/controls insets)}
+   [composer-images/images-list images]
+   [rn/view {:style {:flex-direction :row :margin-top 12}}
+    [image-button chat-id]
+    [rn/view {:width 12}]
+    [reactions-button]
+    [rn/view {:flex 1}]
```
Comment (Parveshdhull) [inline src/status_im2/contexts/chat/messages/composer/controls/view.cljs:70]: Hi, do we need all these overlapping views? Can we just pass these styles as style parameter of these button?

File: src/status_im2/contexts/chat/messages/composer/view.cljs [hunk 5]
```diff
@@ -0,0 +1,250 @@
...
+                        (reanimated/set-shared-value bg-bottom (if (= value 1) 0 (- window-height)))
+                        (reanimated/set-shared-value bg-opacity (reanimated/with-timing value)))
+   :set-translate-y   (fn [value]
+                        (reanimated/set-shared-value translate-y (reanimated/with-timing value)))
+   :set-parent-height (fn [value]
+                        (reanimated/set-shared-value parent-height (reanimated/with-timing value)))})
+
+;; IMPORTANT! PLEASE READ BEFORE MAKING CHANGES
+
+;; 1) insets are not stable so we can't rely on the first values and we have to use it from render
```
Comment (Parveshdhull) [inline src/status_im2/contexts/chat/messages/composer/view.cljs:159]: Yes, that's annoying. But can we just store them, once we get reliable value. I am doing something like that for shell height (as we can't get screen height reliably)
https://github.com/status-im/status-mobile/blob/220341e0be21765380270a579294162580facd10/src/status_im2/contexts/shell/view.cljs#L121
https://github.com/status-im/status-mobile/pull/14639#issuecomment-1366992260
https://github.com/status-im/status-mobile/pull/14639#issuecomment-1366995541
Perhaps we should calculate all screen parameters, (height, width, top, etc) correctly once, and just store them.

**Review by Parveshdhull — APPROVED** [review]

File: src/status_im2/contexts/chat/messages/composer/view.cljs (hunk 5 above)
Comment (flexsurfer) [reply src/status_im2/contexts/chat/messages/composer/view.cljs:159]:
```
[SafeAreaView](https://github.com/th3rdwave/react-native-safe-area-context#safeareaview) is the preferred way to consume insets. This is a regular View with insets applied as extra padding or margin. It offers better performance by applying insets natively and avoids flickers that can happen with the other JS based consumers

[useSafeAreaInsets](https://github.com/th3rdwave/react-native-safe-area-context#usesafeareainsets) offers more flexibility, but can cause some layout flicker in certain cases. Use this if you need more control over how insets are applied.
```

probably we could look at how SafeAreaView is implemented and just implement this in our native code

**Comment by qoqobolo** [comment]
Hi @flexsurfer, thanx for PR! Let's squeeze it :)

### ISSUE 1: White status bar when the Jump to screen is open (iOS and Android)

The bar also remains white after putting the app in the background.

Develop:

Screenshot 2023-01-12 at 11 41 12


PR:
Screenshot 2023-01-12 at 11 41 26

**Review by ibrkhalil — APPROVED** [review]

File: src/quo2/components/avatars/user_avatar.cljs (hunk 2 above)
Comment (flexsurfer) [reply src/quo2/components/avatars/user_avatar.cljs:137]: yes expected

**Comment by qoqobolo** [comment]
### ISSUE 2: The main screen overlaps tabs (Android and iOS))

STR:
1. Create account
2. Tap any tab from the `Jump to` screen

https://user-images.githubusercontent.com/67952253/212049655-1039f659-d833-47a5-9a5e-9aa486a92dac.mp4

Screenshot 2023-01-12 at 12 02 20

**Comment by flexsurfer** [comment]
@Parveshdhull could you please take a look, after rebasing on latest develop, i have these issues in my PR

**Comment by qoqobolo** [comment]
### ISSUE 3: The `Send` button is displayed even without any text in the composer

Design: https://www.figma.com/file/wA8Epdki2OWa8Vr067PCNQ/Composer-for-Mobile?node-id=3101%3A347945&t=7LF19oXbfji6Q5CE-4

Screenshot 2023-01-12 at 12 24 41

**Comment by qoqobolo** [comment]
### ISSUE 4: The `Send` button is not displayed when an image is selected for sending + the `Cross` button is cut off on the image

Screenshot 2023-01-12 at 12 29 53

File: src/status_im2/navigation/view.cljs [hunk 6]
```diff
@@ -38,19 +37,14 @@
    (concat screens/components)))
 
 (defn wrapped-screen-style
-  [{:keys [insets style]} insets-obj]
+  [screen-insets safe-insets]
```
Comment (flexsurfer) [inline src/status_im2/navigation/view.cljs:40]: found it @Parveshdhull, thanks

**Comment by flexsurfer** [comment]
thank @qoqobolo all should be fixed ,except icon cut off, could you give more details, what device etc

**Comment by flexsurfer** [comment]
@qoqobolo pls try , i made changes for image close button

**Comment by VolodLytvynenko** [comment]
Hi @flexsurfer thank you for your work. Issues 1 -4 are fixed. Take a look at additional issues. 
Thanx!

**ISSUE 5: The pin banner is overlapped by messages**

**Steps to reproduce:**
1. Fill the chat with long message
2. Pin the current message

**Actual result:**
The pin is overlapped

https://user-images.githubusercontent.com/52490791/212283977-c13ae067-7291-40db-929d-b28ef8d38255.mp4

**Expected result:**
The pin banner is shown with the text

**Comment by VolodLytvynenko** [comment]
ISSUE 6: The textbox is not expanded fully if the 'edit' option is tapped on the long message
**Steps to reproduce:**

1. Send a long message in the chat
2. Long tap on the message
3. Tap 'edit' option

**Actual result:**
Only one line is shown within the textbox
https://user-images.githubusercontent.com/52490791/212291748-9c6c569c-4399-4dd1-963a-82ddf05f8064.mp4

**Expected result:**
At least 3 lines are shown within the expanded textbox and the cursor is located at the last char

**Env:** Iphone 11 pro v12; Huavei p20 light v9

**Comment by VolodLytvynenko** [comment]
**ISSUE 7. The text is not saved into the textbox if during message editing user navigates out of the chat, then reopens it**

**Steps to reproduce:**
1. Send a message in the chat
2. Long tap on the message -> select 'edit' option
3. Tap 'back' button to navigate out of the chat
4. Open chat again

**Actual result:**
The text is not saved into textbox after chat is reopened
https://user-images.githubusercontent.com/52490791/212293026-4d72d26d-2684-4d2c-a964-58833168ae9f.mp4


**Expected result:**
The text is saved into the textbox after chat is reopened

**Env:** Only IOS. In my case iPhone 11 PRO

**Comment by qoqobolo** [comment]
### ISSUE 8: No shadow behind the drawer on Android

Screenshot 2023-01-13 at 11 07 53

**Comment by flexsurfer** [comment]
ISSUE 5: fixed, most likely it was introduced here https://github.com/status-im/status-mobile/pull/14629

**Comment by flexsurfer** [comment]
ISSUE 6 and ISSUE 7 fixed

**Comment by qoqobolo** [comment]
### ISSUE 9: You can select a mention twice in a multiline message when editing it

This odd behavior is also present in develop. Please let me know if this is out of scope of this PR

**Steps:**
1. Send at least a two-line message
2. Edit the message > Remove the top line (or any other than the bottom line if the message contains more than two lines)  > type `@` > select a name > select it again > send the message

https://user-images.githubusercontent.com/67952253/212297210-19810a03-ddd6-47e1-bf9b-269b9937ea51.mp4

**Comment by flexsurfer** [comment]
ISSUE 8 fixed

**Comment by flexsurfer** [comment]
seems like ISSUE 9 is an old one, and might be tricky to fix, lets file a separate issue, thanks

**Comment by qoqobolo** [comment]
@flexsurfer [ISSUE 6](https://github.com/status-im/status-mobile/pull/14758#issuecomment-1381573162) is still reproducible for me in most cases on Google Pixel 2 (on Huawei it works fine).
Can you re-check, please

https://user-images.githubusercontent.com/67952253/212326990-812e0eff-1a9d-4128-8de0-31e688b08995.mp4

**Comment by flexsurfer** [comment]
@qoqobolo 
image

**Comment by qoqobolo** [comment]
### ISSUE 10: Composer is overlapped by the keyboard when pasting a long message if there is a pinned message in the chat (iOS and Android)

**Steps:**
1. Pin a message in the chat
2. Paste a long message to the composer

https://user-images.githubusercontent.com/67952253/212330362-f2b79ff8-8d0a-4410-8b11-b88dbf16e9c1.mp4

**Comment by flexsurfer** [comment]
thanks @qoqobolo issue10 fixed, issues9 can't replicate on pixel2

**Comment by qoqobolo** [comment]
Thanks a lot for all the fixes @flexsurfer!
E2e failures are not related to PR, it can be merged.

Screenshot 2023-01-16 at 16 37 56

**Pull request merged**